lib/Sema/TypeCheckType.h

```diff
@@ -66,6 +66,9 @@ enum class TypeResolutionFlags : uint16_t {
 
   /// Whether we should not produce diagnostics if the type is invalid.
   SilenceErrors = 1 << 10,
+
+  /// Whether to allow module declaration types.
+  AllowModule = 1 << 11
```

Comment:
@brentdax: I added you as a reviewer because this ad-hoc support for module-qualified and type-qualified declaration names for `@derivative` and `@transpose` attributes overlaps with your work on module-qualified names.

Some context: the main "qualified name" use case for `@derivative` and `@transpose` is module-qualified names. Type-qualified names are no longer necessary after a `@transpose` type-checking rules revamp (TF-1060).

After syntax exists for your work on module-qualified names (e.g. `Module::fooName`), we can drop our ad-hoc support and switch to your work!